programs/stake/src/stake_state.rs

```diff
@@ -1025,49 +1037,44 @@ impl<'a> StakeAccount for KeyedAccount<'a> {
 
     fn merge(
         &self,
-        source_stake: &KeyedAccount,
+        source_account: &KeyedAccount,
         clock: &Clock,
         stake_history: &StakeHistory,
         signers: &HashSet<Pubkey>,
     ) -> Result<(), InstructionError> {
-        if source_stake.owner()? != id() {
+        if source_account.owner()? != id() {
             return Err(InstructionError::IncorrectProgramId);
         }
 
-        let meta = match self.state()? {
-            StakeState::Stake(meta, stake) => {
-                // stake must be fully de-activated
-                if stake.stake(clock.epoch, Some(stake_history), true) != 0 {
-                    return Err(StakeError::MergeActivatedStake.into());
-                }
-                meta
-            }
-            StakeState::Initialized(meta) => meta,
-            _ => return Err(InstructionError::InvalidAccountData),
-        };
+        let (meta, stake) = get_info_if_mergable(self, clock, stake_history)?;
+
         // Authorized staker is allowed to split/merge accounts
         meta.authorized.check(signers, StakeAuthorize::Staker)?;
 
-        let source_meta = match source_stake.state()? {
-            StakeState::Stake(meta, stake) => {
-                // stake must be fully de-activated
-                if stake.stake(clock.epoch, Some(stake_history), true) != 0 {
-                    return Err(StakeError::MergeActivatedStake.into());
-                }
-                meta
-            }
-            StakeState::Initialized(meta) => meta,
-            _ => return Err(InstructionError::InvalidAccountData),
-        };
+        let (source_meta, source_stake) =
+            get_info_if_mergable(source_account, clock, stake_history)?;
 
         // Meta must match for both accounts
         if meta != source_meta {
             return Err(StakeError::MergeMismatch.into());
         }
 
+        // If stake is active, the accounts must be compatible
+        if let Some((mut stake, mut source_stake)) = stake.zip(source_stake) {
```

Comment:
`zip()` is the wrong tool for the job here... It's an error if stake activity is mismatched. `match` instead.